firedrake/preconditioners/fdm.py

```diff
@@ -136,73 +170,919 @@ def interp_nullspace(I, nsp):
+        rindices = None
+        cindices = None
+        if A.getType() != PETSc.Mat.Type.PREALLOCATOR:
+            Ae = self.work_mats[Vrow, Vcol]
+            De = self.work_mats[common_key]
+            data = self.work_csr[2]
+            insert = PETSc.InsertMode.INSERT
+            work_vec = De.getDiagonal()
+            if len(data.shape) == 3:
+                @PETSc.Log.EventDecorator("FDMUpdateDiag")
+                def update_De(data):
+                    De.setValuesCSR(*self.work_csr, addv=insert)
+                    De.assemble()
+                    return De
+            else:
+                @PETSc.Log.EventDecorator("FDMUpdateDiag")
+                def update_De(data):
+                    work_vec.setArray(data)
+                    De.setDiagonal(work_vec, addv=insert)
+                    return De
+
+            for e in range(self.nel):
+                rindices = get_rindices(e, result=rindices)
+                cindices = get_cindices(e, result=cindices)
+                data = self.get_coeffs(e, result=data)
+                Ae = assemble_element_mat(update_De(data), result=Ae)
+                update_A(condense_element_mat(Ae), rindices, cindices)
+
+            work_vec.destroy()
+
+        elif self.nel:
+            if common_key not in self.work_mats:
+                data = self.get_coeffs(0)
+                data.fill(1.0E0)
+                shape = data.shape + (1,)*(3-len(data.shape))
+                nrows = shape[0] * shape[1]
+                ai = numpy.arange(nrows+1, dtype=PETSc.IntType)
+                aj = numpy.tile(ai[:-1].reshape((-1, shape[1])), (1, shape[2]))
+                if shape[2] > 1:
+                    ai *= shape[2]
+                    data = numpy.tile(numpy.eye(shape[2]), shape[:1] + (1,)*(len(shape)-1))
+
+                self.work_csr = (ai, aj, data)
+                De = PETSc.Mat().createAIJ((nrows, nrows), csr=self.work_csr, comm=PETSc.COMM_SELF)
+                self.work_mats[common_key] = De
+
+            De = self.work_mats[common_key]
+            Ae = assemble_element_mat(De, result=None)
+            self.work_mats[Vrow, Vcol] = Ae
+            if do_sort:
+                sort_interior_dofs(self.ises[0], Ae)
+            Se = condense_element_mat(Ae)
+
+            for e in range(self.nel):
+                rindices = get_rindices(e, result=rindices)
+                cindices = get_cindices(e, result=cindices)
+                update_A(Se, rindices, cindices)
+        else:
+            self.work_csr = (None, None, None)
+            self.work_mats[common_key] = None
+            self.work_mats[Vrow, Vcol] = None
+        if RtAP.buff:
+            RtAP.buff.destroy()
+
+    @PETSc.Log.EventDecorator("FDMCoefficients")
+    def assemble_coef(self, J, form_compiler_parameters):
         """
-        Assemble the sparse preconditioner with cell-wise constant coefficients.
+        Obtain coefficients as the diagonal of a weighted mass matrix in V^k x V^{k+1}
```

Comment:
Refer to discussion in manuscript.

Explain basic strategy for how these will be extracted.